exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -514,29 +517,22 @@ func getNewMachine(mp *expv1.MachinePool, infraMachine *unstructured.Unstructure
 func (r *MachinePoolReconciler) infraMachineToMachinePoolMapper(ctx context.Context, o client.Object) []ctrl.Request {
 	log := ctrl.LoggerFrom(ctx)
 
-	machinePoolList := &expv1.MachinePoolList{}
 	labels := o.GetLabels()
-	selector := map[string]string{}
-	if clusterName, ok := labels[clusterv1.ClusterNameLabel]; ok {
-		selector = map[string]string{clusterv1.ClusterNameLabel: clusterName}
-	}
-
-	if poolNameHash, ok := labels[clusterv1.MachinePoolNameLabel]; ok {
-		if err := r.Client.List(ctx, machinePoolList, client.InNamespace(o.GetNamespace()), client.MatchingLabels(selector)); err != nil {
-			log.Error(err, "failed to get MachinePool for InfraMachine")
+	_, machinePoolOwned := labels[clusterv1.MachinePoolNameLabel]
+	if machinePoolOwned {
+		machinePool, err := utilexp.GetMachinePoolByLabels(ctx, r.Client, o.GetNamespace(), labels)
```

Comment:
The issue here is that the value of `clusterv1.MachinePoolNameLabel` could be a hash so we don't know the name of the MachinePool. In `GetMachinePoolByLabels()`, we're listing the MachinePools and checking to see if their name would hash to the same value as the label and returning the name of the MP that matches. I'm not sure if we can find the name of the MP if it doesn't exist.

Reply:
Is this design smell in the contents of this label? Labels are typically used for efficient selection of objects, but this label can't be used this way. Do we need a new label?

I don't know the reason why the label contains a hash of the name, but in order to make selection efficient I think we would need to either:
* Add a label to the MachinePool containing the hashed value, or
* Add a new label containing the un-hashed name to the Machine

Reply:
IIRC the issue was that if the label gets too long, it gets converted to a hash since the name of a K8s object can be longer than the max label length. I think adding the hashed value to the MachinePool would be reasonable, although it would look a little odd.